<a href="https://colab.research.google.com/github/ArtemFilosofGB/PythonForAnalyst/blob/main/Lesson_2_%D0%90%D0%BD%D0%B0%D0%BB%D0%B8%D0%B7_%D0%B4%D0%B0%D0%BD%D0%BD%D1%8B%D1%85_%D1%81_%D0%B1%D0%B8%D0%B1%D0%BB%D0%B8%D0%BE%D1%82%D0%B5%D0%BA%D0%BE%D0%B9_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Урок 2. Анализ данных с библиотекой Pandas

In [19]:
import pandas as pd
import numpy as np

Series
Она представляет из себя объект, похожий на одномерный массив, но отличительной чертой является наличие индексов. Индекс находится слева, а сам элемент справа.

Синтаксис создания:

pandas.Series(input_data, index, data_type)
input_data: ввод в виде списка, константы, массива NumPy, Dict и т. д.
index: значения индексов.
data_type (опционально): тип данных.

In [20]:
a = pd.Series([4, 7, 6, 3, 9],
index=['one', 'two', 'three', 'four', 'five'])
a

one      4
two      7
three    6
four     3
five     9
dtype: int64

In [21]:
a = pd.Series([4, 7, 6, 3, 9])
a

0    4
1    7
2    6
3    3
4    9
dtype: int64

In [22]:
a.index

RangeIndex(start=0, stop=5, step=1)

In [23]:
a.values

array([4, 7, 6, 3, 9], dtype=int64)

In [24]:
a[0]

4

In [25]:
a['one']

KeyError: 'one'

DataFrame
Объект DataFrame является табличной структурой данных. В любой таблице всегда присутствуют строки и столбцы. При этом в столбцах можно хранить данные разных типов данных. Столбцами в объекте DataFrame выступают объекты Series, строки которых являются их элементами.

Синтаксис создания:

pandas.DataFrame(input_data, index)
input_data: ввод в виде Dict, 2D массива NumPy, Series и т. д.
index: значения индексов.

In [ ]:
df = pd.DataFrame({
    'Age': [46, 37, 44, 42, 42],
    'Country': ['Spain', 'Spain', 'Germany', 'Germany', 'France'],
    'Gender': ['Female', 'Female', 'Male', 'Male', 'Male']
})

df

Age  Country  Gender
0   46    Spain  Female
1   37    Spain  Female
2   44  Germany    Male
3   42  Germany    Male
4   42   France    Male

In [ ]:
df['Age']

0    46
1    37
2    44
3    42
4    42
Name: Age, dtype: int64

In [ ]:
df.Country

0      Spain
1      Spain
2    Germany
3    Germany
4     France
Name: Country, dtype: object

In [ ]:
df[['Country', 'Age']]

Country  Age
0    Spain   46
1    Spain   37
2  Germany   44
3  Germany   42
4   France   42

df.index

In [ ]:
df = pd.DataFrame({
    'Age': [46, 37, 44, 42, 42],
    'Country': ['Spain', 'Spain', 'Germany', 'Germany', 'France'],
    'Gender': ['Female', 'Female', 'Male', 'Male', 'Male']
}, index=[5, 4, 6, 3, 2])

df

Age  Country  Gender
5   46    Spain  Female
4   37    Spain  Female
6   44  Germany    Male
3   42  Germany    Male
2   42   France    Male

In [ ]:
df.index = [101, 102, 103, 104, 105]
df

Age  Country  Gender
101   46    Spain  Female
102   37    Spain  Female
103   44  Germany    Male
104   42  Germany    Male
105   42   France    Male

#Считывание данных
В целом, pandas поддерживает все самые популярные форматы хранения данных: csv, excel, sql, html и многое другое, но чаще всего приходится работать именно с csv файлами (comma separated values).

Будем работать с датасетом по оттоку клиентов из банка https://www.kaggle.com/datasets/shubh0799/churn-modelling.

Характеристики каждого клиента:

RowNumber - Номер строки

CustomerId - Уникальный идентификатор клиента

CreditScore - Кредитная оценка клиента

Geography - Из какой страны клиент

Gender - Пол клиента

Age - Возраст клиента

Tenure - Сколько лет человек является клиентом банка

Balance - Баланс счета

NumOfProducts - Количество открытых продуктов

HasCrCard - Есть ли у клиента кредитная карта

IsActiveMember - Является ли клиент активные участником

EstimatedSalary - Предположительная зарплата клиента

Exited - Уйдет ли человек в отток





In [ ]:
df = pd.read_csv('Churn_Modelling.csv')
df


RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0          2       0.00              1          1               1   
1          1   83807.86              1          0               1   
2          8  159660.80              3          1               0   
3          1       0.00              2          0               0   
4          2  125510.82              1          1               1   
...      ...        ...            ...        ...             ...   
9995       5       0.00              2          1               0   
9996      10   57369.61              1          1               1   
9997       7       0.00              1          0               1   
9998       3   75075.31              2          1               0   
9999       4  130142.79              1          1               0   

      EstimatedSalary  Exited  
0           101348.88       1  
1           112542.58       0  
2           113931.57       1  
3            93826.63       0  
4            79084.10       0  
...               ...     ...  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0  

[10000 rows x 14 columns]

In [ ]:
pd.read_csv('./Churn_Modelling.csv', header=1)


1  15634602   Hargrave  619   France  Female  42   2          0  \
0         2  15647311       Hill  608    Spain  Female  41   1   83807.86   
1         3  15619304       Onio  502   France  Female  42   8  159660.80   
2         4  15701354       Boni  699   France  Female  39   1       0.00   
3         5  15737888   Mitchell  850    Spain  Female  43   2  125510.82   
4         6  15574012        Chu  645    Spain    Male  44   8  113755.78   
...     ...       ...        ...  ...      ...     ...  ..  ..        ...   
9994   9996  15606229   Obijiaku  771   France    Male  39   5       0.00   
9995   9997  15569892  Johnstone  516   France    Male  35  10   57369.61   
9996   9998  15584532        Liu  709   France  Female  36   7       0.00   
9997   9999  15682355  Sabbatini  772  Germany    Male  42   3   75075.31   
9998  10000  15628319     Walker  792   France  Female  28   4  130142.79   

      1.1  1.2  1.3  101348.88  1.4  
0       1    0    1  112542.58    0  
1       3    1    0  113931.57    1  
2       2    0    0   93826.63    0  
3       1    1    1   79084.10    0  
4       2    1    0  149756.71    1  
...   ...  ...  ...        ...  ...  
9994    2    1    0   96270.64    0  
9995    1    1    1  101699.77    0  
9996    1    0    1   42085.58    1  
9997    2    1    0   92888.52    1  
9998    1    1    0   38190.78    0  

[9999 rows x 14 columns]

In [ ]:
pd.read_csv('./Churn_Modelling.csv', sep=';')


RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
0     1,15634602,Hargrave,619,France,Female,42,2,0,1...                                                                                        
1     2,15647311,Hill,608,Spain,Female,41,1,83807.86...                                                                                        
2     3,15619304,Onio,502,France,Female,42,8,159660....                                                                                        
3     4,15701354,Boni,699,France,Female,39,1,0,2,0,0...                                                                                        
4     5,15737888,Mitchell,850,Spain,Female,43,2,1255...                                                                                        
...                                                 ...                                                                                        
9995  9996,15606229,Obijiaku,771,France,Male,39,5,0,...                                                                                        
9996  9997,15569892,Johnstone,516,France,Male,35,10,...                                                                                        
9997  9998,15584532,Liu,709,France,Female,36,7,0,1,0...                                                                                        
9998  9999,15682355,Sabbatini,772,Germany,Male,42,3,...                                                                                        
9999  10000,15628319,Walker,792,France,Female,28,4,1...                                                                                        

[10000 rows x 1 columns]

In [ ]:
df

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0          2       0.00              1          1               1   
1          1   83807.86              1          0               1   
2          8  159660.80              3          1               0   
3          1       0.00              2          0               0   
4          2  125510.82              1          1               1   
...      ...        ...            ...        ...             ...   
9995       5       0.00              2          1               0   
9996      10   57369.61              1          1               1   
9997       7       0.00              1          0               1   
9998       3   75075.31              2          1               0   
9999       4  130142.79              1          1               0   

      EstimatedSalary  Exited  
0           101348.88       1  
1           112542.58       0  
2           113931.57       1  
3            93826.63       0  
4            79084.10       0  
...               ...     ...  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0  

[10000 rows x 14 columns]

In [ ]:
df.head()


RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

df.tail()

In [ ]:
df.tail()

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
9995       5       0.00              2          1               0   
9996      10   57369.61              1          1               1   
9997       7       0.00              1          0               1   
9998       3   75075.31              2          1               0   
9999       4  130142.79              1          1               0   

      EstimatedSalary  Exited  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0

In [ ]:
df.sample(frac=1) #перемешанный порядок

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
8841       8842    15718769    Fallaci          557     Spain    Male   36   
8851       8852    15590828   Chidimma          782   Germany    Male   42   
5827       5828    15801701     Robson          653     Spain    Male   35   
6927       6928    15781665     Ibekwe          601    France  Female   37   
271         272    15619955  Bevington          733   Germany    Male   34   
...         ...         ...        ...          ...       ...     ...  ...   
2942       2943    15715541       Yang          850    France  Female   42   
3632       3633    15684696        Lei          560     Spain  Female   26   
3510       3511    15800814    Palerma          534    France    Male   35   
6648       6649    15731751   Osinachi          437    France  Female   26   
9969       9970    15806360        Hou          609    France    Male   41   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
8841       1  113110.26              1          1               0   
8851       7  126428.38              1          1               0   
5827       9       0.00              2          1               1   
6927       5       0.00              1          0               0   
271        3  100337.96              3          1               0   
...      ...        ...            ...        ...             ...   
2942       9  113311.11              1          1               1   
3632       3  116576.45              1          1               0   
3510       2   81951.74              2          1               0   
6648       1  120923.52              1          0               1   
9969       6       0.00              1          0               1   

      EstimatedSalary  Exited  
8841         98413.10       0  
8851         39830.10       0  
5827         45956.05       0  
6927         20708.60       0  
271          48559.19       1  
...               ...     ...  
2942        198193.75       0  
3632        157567.37       0  
3510        115668.53       0  
6648         78854.57       0  
9969        112585.19       0  

[10000 rows x 14 columns]

In [ ]:
df.shape


(10000, 14)

# Первичный анализ данных
Типы данных:

int: целочисленные значения. Пример: 9, 56, 30
float: вещественные значения (с плавающей точкой). Пример: 7.3, 9.0, 45.334
object/str: строковые значения. Пример: ‘hello, world’, ‘50 000’
df.info()


<class 'pandas.core.frame.DataFrame'>


RangeIndex: 10000 entries, 0 to 9999

Data columns (total 14 columns):

RowNumber          10000 non-null int64

CustomerId         10000 non-null int64

Surname            10000 non-null object

CreditScore        10000 non-null int64

Geography          10000 non-null object

Gender             10000 non-null object

Age                10000 non-null int64

Tenure             10000 non-null int64

Balance            10000 non-null float64

NumOfProducts      10000 non-null int64

HasCrCard          10000 non-null int64

IsActiveMember     10000 non-null int64

EstimatedSalary    10000 non-null float64

Exited             10000 non-null int64

dtypes: float64(2), int64(9), object(3)

memory usage: 1.1+ MB

df.describe()

Выводятся значения:

Count - количество непропущенных объектов (там, где нет nan значений)

mean - арифметическое среднее

std - стандартное отклонение

min - минимальное значение

25% - квантиль 25 процентов

50% - квантиль 50 процентов или же медиана

75% - квантиль 75 процентов

max - максимальное значение


In [ ]:
df.describe() # не возвращаются строковые и bool


RowNumber    CustomerId   CreditScore           Age        Tenure  \
count  10000.00000  1.000000e+04  10000.000000  10000.000000  10000.000000   
mean    5000.50000  1.569094e+07    650.528800     38.921800      5.012800   
std     2886.89568  7.193619e+04     96.653299     10.487806      2.892174   
min        1.00000  1.556570e+07    350.000000     18.000000      0.000000   
25%     2500.75000  1.562853e+07    584.000000     32.000000      3.000000   
50%     5000.50000  1.569074e+07    652.000000     37.000000      5.000000   
75%     7500.25000  1.575323e+07    718.000000     44.000000      7.000000   
max    10000.00000  1.581569e+07    850.000000     92.000000     10.000000   

             Balance  NumOfProducts    HasCrCard  IsActiveMember  \
count   10000.000000   10000.000000  10000.00000    10000.000000   
mean    76485.889288       1.530200      0.70550        0.515100   
std     62397.405202       0.581654      0.45584        0.499797   
min         0.000000       1.000000      0.00000        0.000000   
25%         0.000000       1.000000      0.00000        0.000000   
50%     97198.540000       1.000000      1.00000        1.000000   
75%    127644.240000       2.000000      1.00000        1.000000   
max    250898.090000       4.000000      1.00000        1.000000   

       EstimatedSalary        Exited  
count     10000.000000  10000.000000  
mean     100090.239881      0.203700  
std       57510.492818      0.402769  
min          11.580000      0.000000  
25%       51002.110000      0.000000  
50%      100193.915000      0.000000  
75%      149388.247500      0.000000  
max      199992.480000      1.000000

In [ ]:
df['Balance'].max()

250898.09

In [ ]:
print(df['Age'].max())
print(df['Age'].mean())
print(df['Age'].min())

92
38.9218
18


In [ ]:
df[['CreditScore','Age','Tenure']].mean()

CreditScore    650.5288
Age             38.9218
Tenure           5.0128
dtype: float64

In [ ]:
df.describe(include=['object'])

Surname Geography Gender
count    10000     10000  10000
unique    2932         3      2
top      Smith    France   Male
freq        32      5014   5457

In [ ]:
df.dtypes # узнать типы полей таблицы


RowNumber            int64
CustomerId           int64
Surname             object
CreditScore          int64
Geography           object
Gender              object
Age                  int64
Tenure               int64
Balance            float64
NumOfProducts        int64
HasCrCard            int64
IsActiveMember       int64
EstimatedSalary    float64
Exited               int64
dtype: object

In [ ]:
df['Age'].dtype # если серия исп type

dtype('int64')

In [ ]:
df['HasCrCard'].astype('bool') #смена при отображении типа у столбца с int64 на bool

0        True
1       False
2        True
3       False
4        True
        ...  
9995     True
9996     True
9997    False
9998     True
9999     True
Name: HasCrCard, Length: 10000, dtype: bool

In [ ]:
df['HasCrCard'].dtype

dtype('int64')

In [ ]:
df['HasCrCard']=df['HasCrCard'].astype('bool') #смена типа у столбца с int64 на bool

In [ ]:
df['HasCrCard'].dtype


dtype('bool')

In [ ]:
df['Geography'].unique()


array(['France', 'Spain', 'Germany'], dtype=object)

In [ ]:
df['Geography'].value_counts()

France     5014
Germany    2509
Spain      2477
Name: Geography, dtype: int64

In [ ]:
df['Geography'].value_counts(normalize=True)

France     0.5014
Germany    0.2509
Spain      0.2477
Name: Geography, dtype: float64

# Фильтрация
Фильтрация в pandas основывается на булевых масках.

Булевая маска — бинарные данные, которые используются для выбора определенных объектов из структуры данных.

In [ ]:
male = df['Gender'] == 'Male'

0       False
1       False
2       False
3       False
4       False
        ...  
9995     True
9996     True
9997    False
9998     True
9999    False
Name: Gender, Length: 10000, dtype: bool

In [ ]:
male = df[df['Gender'] == 'Male']
male

RowNumber  CustomerId      Surname  CreditScore Geography Gender  Age  \
5             6    15574012          Chu          645     Spain   Male   44   
6             7    15592531     Bartlett          822    France   Male   50   
8             9    15792365           He          501    France   Male   44   
9            10    15592389           H?          684    France   Male   27   
10           11    15767821       Bearce          528    France   Male   31   
...         ...         ...          ...          ...       ...    ...  ...   
9992       9993    15657105  Chukwualuka          726     Spain   Male   36   
9993       9994    15569266       Rahman          644    France   Male   28   
9995       9996    15606229     Obijiaku          771    France   Male   39   
9996       9997    15569892    Johnstone          516    France   Male   35   
9998       9999    15682355    Sabbatini          772   Germany   Male   42   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
5          8  113755.78              2       True               0   
6          7       0.00              2       True               1   
8          4  142051.07              2      False               1   
9          2  134603.88              1       True               1   
10         6  102016.72              2      False               0   
...      ...        ...            ...        ...             ...   
9992       2       0.00              1       True               0   
9993       7  155060.41              1       True               0   
9995       5       0.00              2       True               0   
9996      10   57369.61              1       True               1   
9998       3   75075.31              2       True               0   

      EstimatedSalary  Exited  
5           149756.71       1  
6            10062.80       0  
8            74940.50       0  
9            71725.73       0  
10           80181.12       0  
...               ...     ...  
9992        195192.40       0  
9993         29179.52       0  
9995         96270.64       0  
9996        101699.77       0  
9998         92888.52       1  

[5457 rows x 14 columns]

# Логические И
При операторе & нужно, чтобы выполнялось два условия одновременно:

In [ ]:
df[(df['Gender'] == 'Female') & (df['NumOfProducts'] >= 3)]

RowNumber  CustomerId        Surname  CreditScore Geography  Gender  \
2             3    15619304           Onio          502    France  Female   
7             8    15656148         Obinna          376   Germany  Female   
30           31    15589475        Azikiwe          591     Spain  Female   
88           89    15622897         Sharpe          646    France  Female   
90           91    15757535           Heap          647     Spain  Female   
...         ...         ...            ...          ...       ...     ...   
9565       9566    15752294           Long          582    France  Female   
9747       9748    15775761  Iweobiegbunam          610   Germany  Female   
9800       9801    15640507             Li          762     Spain  Female   
9877       9878    15572182     Onwuamaeze          505   Germany  Female   
9895       9896    15796764          Bruno          684   Germany  Female   

      Age  Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
2      42       8  159660.80              3       True               0   
7      29       4  115046.74              4       True               0   
30     39       3       0.00              3       True               0   
88     46       4       0.00              3       True               0   
90     44       5       0.00              3       True               1   
...   ...     ...        ...            ...        ...             ...   
9565   38       9  135979.01              4       True               1   
9747   69       5   86038.21              3      False               0   
9800   35       3  119349.69              3       True               1   
9877   33       3  106506.77              3       True               0   
9895   56       3  127585.98              3       True               1   

      EstimatedSalary  Exited  
2           113931.57       1  
7           119346.88       1  
30          140469.38       1  
88           93251.42       1  
90          174205.22       1  
...               ...     ...  
9565         76582.95       1  
9747        192743.06       1  
9800         47114.18       1  
9877         45445.78       1  
9895         80593.49       1  

[187 rows x 14 columns]

# Логические ИЛИ
При операторе | нужно, чтобы выполнялось хотя бы одно условие:

In [ ]:
df[(df['HasCrCard']) | (df['NumOfProducts'] >= 3)]

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
2             3    15619304       Onio          502    France  Female   42   
4             5    15737888   Mitchell          850     Spain  Female   43   
5             6    15574012        Chu          645     Spain    Male   44   
6             7    15592531   Bartlett          822    France    Male   50   
...         ...         ...        ...          ...       ...     ...  ...   
9993       9994    15569266     Rahman          644    France    Male   28   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0          2       0.00              1       True               1   
2          8  159660.80              3       True               0   
4          2  125510.82              1       True               1   
5          8  113755.78              2       True               0   
6          7       0.00              2       True               1   
...      ...        ...            ...        ...             ...   
9993       7  155060.41              1       True               0   
9995       5       0.00              2       True               0   
9996      10   57369.61              1       True               1   
9998       3   75075.31              2       True               0   
9999       4  130142.79              1       True               0   

      EstimatedSalary  Exited  
0           101348.88       1  
2           113931.57       1  
4            79084.10       0  
5           149756.71       1  
6            10062.80       0  
...               ...     ...  
9993         29179.52       0  
9995         96270.64       0  
9996        101699.77       0  
9998         92888.52       1  
9999         38190.78       0  

[7150 rows x 14 columns]

# Логические НЕ ~
При операторе ~ булевая маска обращается: True меняется на False и наоборот:

In [ ]:
df[~(df['Geography'] == 'Spain')]

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
6             7    15592531   Bartlett          822    France    Male   50   
7             8    15656148     Obinna          376   Germany  Female   29   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0          2       0.00              1       True               1   
2          8  159660.80              3       True               0   
3          1       0.00              2      False               0   
6          7       0.00              2       True               1   
7          4  115046.74              4       True               0   
...      ...        ...            ...        ...             ...   
9995       5       0.00              2       True               0   
9996      10   57369.61              1       True               1   
9997       7       0.00              1      False               1   
9998       3   75075.31              2       True               0   
9999       4  130142.79              1       True               0   

      EstimatedSalary  Exited  
0           101348.88       1  
2           113931.57       1  
3            93826.63       0  
6            10062.80       0  
7           119346.88       1  
...               ...     ...  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0  

[7523 rows x 14 columns]

In [ ]:
df[df['Geography'].isin(['France', 'Germany'])]

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
6             7    15592531   Bartlett          822    France    Male   50   
7             8    15656148     Obinna          376   Germany  Female   29   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0          2       0.00              1       True               1   
2          8  159660.80              3       True               0   
3          1       0.00              2      False               0   
6          7       0.00              2       True               1   
7          4  115046.74              4       True               0   
...      ...        ...            ...        ...             ...   
9995       5       0.00              2       True               0   
9996      10   57369.61              1       True               1   
9997       7       0.00              1      False               1   
9998       3   75075.31              2       True               0   
9999       4  130142.79              1       True               0   

      EstimatedSalary  Exited  
0           101348.88       1  
2           113931.57       1  
3            93826.63       0  
6            10062.80       0  
7           119346.88       1  
...               ...     ...  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0  

[7523 rows x 14 columns]

# Индексация

In [ ]:
df_small = df[(df['Geography'] == 'Spain')][['Geography', 'Gender', 'Age']]
df_small.head()

Geography  Gender  Age
1      Spain  Female   41
4      Spain  Female   43
5      Spain    Male   44
11     Spain    Male   24
14     Spain  Female   35

# loc

In [ ]:
df_small.loc[1] #1 индекс

Geography     Spain
Gender       Female
Age              41
Name: 1, dtype: object

In [ ]:
df_small.loc[11]

Geography    Spain
Gender        Male
Age             24
Name: 11, dtype: object

In [ ]:
df_small.iloc[0] #обращение по порядковому индексу

Geography     Spain
Gender       Female
Age              41
Name: 1, dtype: object

# Сортировки

In [ ]:
df.sort_values('Age',ascending=False) # по убыыванию

RowNumber  CustomerId      Surname  CreditScore Geography  Gender  Age  \
6443       6444    15764927       Rogova          753    France    Male   92   
6759       6760    15660878        T'ien          705    France    Male   92   
2458       2459    15813303      Rearick          513     Spain    Male   88   
3033       3034    15578006          Yao          787    France  Female   85   
3387       3388    15798024         Lori          537   Germany    Male   84   
...         ...         ...          ...          ...       ...     ...  ...   
9782       9783    15728829       Weigel          509    France    Male   18   
2141       2142    15758372      Wallace          674    France    Male   18   
9501       9502    15634146          Hou          835   Germany    Male   18   
9520       9521    15673180  Onyekaozulu          727   Germany  Female   18   
1619       1620    15770309     McDonald          656    France    Male   18   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
6443       3  121513.31              1          0               1   
6759       1  126076.24              2          1               1   
2458      10       0.00              2          1               1   
3033      10       0.00              2          1               1   
3387       8   92242.34              1          1               1   
...      ...        ...            ...        ...             ...   
9782       7  102983.91              1          1               0   
2141       7       0.00              2          1               1   
9501       2  142872.36              1          1               1   
9520       2   93816.70              2          1               0   
1619      10  151762.74              1          0               1   

      EstimatedSalary  Exited  
6443        195563.99       0  
6759         34436.83       0  
2458         52952.24       0  
3033        116537.96       0  
3387        186235.98       0  
...               ...     ...  
9782        171770.58       0  
2141         55753.12       1  
9501        117632.63       0  
9520        126172.11       0  
1619        127014.32       0  

[10000 rows x 14 columns]

In [26]:
df.sort_values('Age') # По возростанию

RowNumber  CustomerId      Surname  CreditScore Geography  Gender  Age  \
3512       3513    15657779       Boylan          806     Spain    Male   18   
1678       1679    15569178    Kharlamov          570    France  Female   18   
3517       3518    15757821      Burgess          771     Spain    Male   18   
9520       9521    15673180  Onyekaozulu          727   Germany  Female   18   
2021       2022    15795519     Vasiliev          716   Germany  Female   18   
...         ...         ...          ...          ...       ...     ...  ...   
3387       3388    15798024         Lori          537   Germany    Male   84   
3033       3034    15578006          Yao          787    France  Female   85   
2458       2459    15813303      Rearick          513     Spain    Male   88   
6759       6760    15660878        T'ien          705    France    Male   92   
6443       6444    15764927       Rogova          753    France    Male   92   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
3512       3       0.00              2          1               1   
1678       4   82767.42              1          1               0   
3517       1       0.00              2          0               0   
9520       2   93816.70              2          1               0   
2021       3  128743.80              1          0               0   
...      ...        ...            ...        ...             ...   
3387       8   92242.34              1          1               1   
3033      10       0.00              2          1               1   
2458      10       0.00              2          1               1   
6759       1  126076.24              2          1               1   
6443       3  121513.31              1          0               1   

      EstimatedSalary  Exited  
3512         86994.54       0  
1678         71811.90       0  
3517         41542.95       0  
9520        126172.11       0  
2021        197322.13       0  
...               ...     ...  
3387        186235.98       0  
3033        116537.96       0  
2458         52952.24       0  
6759         34436.83       0  
6443        195563.99       0  

[10000 rows x 14 columns]

In [28]:
df.sort_values(['Age','CreditScore']) # внутри одинаковых значений Age сортируются по CreditScore

RowNumber  CustomerId          Surname  CreditScore Geography  Gender  \
9782       9783    15728829           Weigel          509    France    Male   
1678       1679    15569178        Kharlamov          570    France  Female   
9029       9030    15722701            Bruno          594   Germany    Male   
7334       7335    15759133          Vaguine          616    France    Male   
9526       9527    15665521  Chiazagomekpele          642   Germany    Male   
...         ...         ...              ...          ...       ...     ...   
3387       3388    15798024             Lori          537   Germany    Male   
3033       3034    15578006              Yao          787    France  Female   
2458       2459    15813303          Rearick          513     Spain    Male   
6759       6760    15660878            T'ien          705    France    Male   
6443       6444    15764927           Rogova          753    France    Male   

      Age  Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
9782   18       7  102983.91              1          1               0   
1678   18       4   82767.42              1          1               0   
9029   18       1  132694.73              1          1               0   
7334   18       6       0.00              2          1               1   
9526   18       5  111183.53              2          0               1   
...   ...     ...        ...            ...        ...             ...   
3387   84       8   92242.34              1          1               1   
3033   85      10       0.00              2          1               1   
2458   88      10       0.00              2          1               1   
6759   92       1  126076.24              2          1               1   
6443   92       3  121513.31              1          0               1   

      EstimatedSalary  Exited  
9782        171770.58       0  
1678         71811.90       0  
9029        167689.56       0  
7334         27308.58       0  
9526         10063.75       0  
...               ...     ...  
3387        186235.98       0  
3033        116537.96       0  
2458         52952.24       0  
6759         34436.83       0  
6443        195563.99       0  

[10000 rows x 14 columns]